base/learner.h

```diff
@@ -0,0 +1,45 @@
+#ifndef  _MACHINELEARNING_BASE_LEARNER_H_ 
+#define  _MACHINELEARNING_BASE_LEARNER_H_
+
+#include <vector>
+
+namespace base {
+
+/**
+ * Base class for all learning algorithm implementations.
+ */
+class Learner{
+
+  public:
+
+    /**
+     * \brief Standard destructor.
+     */
+    virtual ~Learner();
+
+    /**
+     * Train the learner with a given dataset.
+     * \param X A pointer object of Collection.
+     */
+    virtual void Train(/*const Collection *X*/);
+
+    /**
+     * Predicts the labels of the instances of X based on the fitted model.
+     * \param X           A pointer object of Collection.
+     * \param prediction  Instances prediction.
+     */
+    virtual void Predict(/*const Collection *X,*/ std::vector<double> &prediction) = 0;
```

Comment:
I think you mixed pointer and reference here.
Inputs must be passed by reference, and outputs using a pointer. So it should be Collection &X and std::vector<double> *prediction. (The same on  PredictProbability)